src/Features/Core/Portable/EditAndContinue/EditAndContinueWorkspaceService.cs

```diff
@@ -0,0 +1,488 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Internal.Log;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.EditAndContinue
+{
+    /// <summary>
+    /// Implements core of Edit and Continue orchestration: management of edit sessions and connecting EnC related services.
+    /// </summary>
+    internal sealed class EditAndContinueWorkspaceService : IEditAndContinueWorkspaceService
+    {
+        internal static readonly TraceLog Log = new TraceLog(2048, "EnC");
+
+        private readonly Workspace _workspace;
+        private readonly IActiveStatementTrackingService _trackingService;
+        private readonly IActiveStatementProvider _activeStatementProvider;
+        private readonly IDiagnosticAnalyzerService _diagnosticService;
+        private readonly IDebuggeeModuleMetadataProvider _debugeeModuleMetadataProvider;
+        private readonly EditAndContinueDiagnosticUpdateSource _emitDiagnosticsUpdateSource;
+        private readonly EditSessionTelemetry _editSessionTelemetry;
+        private readonly DebuggingSessionTelemetry _debuggingSessionTelemetry;
+
+        /// <summary>
+        /// A document id is added whenever a diagnostic is reported while in run mode.
+        /// These diagnostics are cleared as soon as we enter break mode or the debugging session terminates.
+        /// </summary>
+        private readonly HashSet<DocumentId> _documentsWithReportedDiagnosticsDuringRunMode;
+        private readonly object _documentsWithReportedDiagnosticsDuringRunModeGuard = new object();
+
+        private DebuggingSession _debuggingSession;
+        private EditSession _editSession;
+        private PendingSolutionUpdate _pendingUpdate;
+
+        internal EditAndContinueWorkspaceService(
```

Comment:
```suggestion
        [Obsolete(MefConstruction.FactoryMethodMessage, error: true)]
        internal EditAndContinueWorkspaceService(
```

Reply:
This is used by EditAndContinueWorkspaceServiceFactory and tests to construct the instance.

---
In reply to: [270627662](https://github.com/dotnet/roslyn/pull/34027#discussion_r270627662) [](ancestors = 270627662)

Reply:
➡️ Leaving for follow-up in #35439